doc/reference/trd-hil-design.md

```diff
@@ -0,0 +1,655 @@
+    fn transmit(&self, tx_data: &'static mut [u8], tx_len: usize);
+    fn receive(&self, rx_buffer: &'static mut [u8], rx_len: usize);
+    fn abort_receive(&self);
+}
+```
+
+It breaks both Rule 7 and Rule 8. It couples reception and
+transmission (Rule 7).  It also couples configuration with data (Rule
+8). This HIL was fine when there was only a single user of the
+UART. However, once the UART was virtualized, `configure` could not
+work for virtualized clients. There were two options: have `configure`
+always return an error for virtual clients, or write a new trait for
+virtual clients that did not have `configure`. Neither is a good
+solution. The first pushes failures to runtime: a capsule that needs
+to adjust the configuration of the UART can be connected to a virtual
+UART and compile fine, but then fails when it tries to call
+`configure`. If that occurs rarely, then it might be a long time until
+the problem is discovered. The second solution (a new trait) breaks
+the idea of virtualization: a client has to be bound to either a
+physical UART or a virtual one, and can't be swapped between them even
+if it never calls `configure`.
+
+The modern UART HIL looks like this:
+
+```rust
+pub trait Configure {
+    fn configure(&self, params: Parameters) -> ReturnCode;
+}
+pub trait Transmit<'a> {
+    fn set_transmit_client(&self, client: &'a dyn TransmitClient);
+    fn transmit_buffer(
+        &self,
+        tx_buffer: &'static mut [u8],
+        tx_len: usize,
+    ) -> (ReturnCode, Option<&'static mut [u8]>);
+    fn transmit_word(&self, word: u32) -> ReturnCode;
+    fn transmit_abort(&self) -> ReturnCode;
+}
+pub trait Receive<'a> {
+    fn set_receive_client(&self, client: &'a dyn ReceiveClient);
+    fn receive_buffer(
+        &self,
+        rx_buffer: &'static mut [u8],
+        rx_len: usize,
+    ) -> (ReturnCode, Option<&'static mut [u8]>);
+    fn receive_word(&self) -> ReturnCode;
+    fn receive_abort(&self) -> ReturnCode;
+}
+pub trait Uart<'a>: Configure + Transmit<'a> + Receive<'a> {}
+pub trait UartData<'a>: Transmit<'a> + Receive<'a> {}
+```
+
+Rule 9: Avoid Blocking APIs
+===============================
+
+The Tock kernel is non-blocking: I/O operations are split-phase and
+have a completion callback. If an operation blocks, it blocks the
+entire system.
+
+There are cases when operations are synchronous *sometimes*. The
+random number generator in Rule 1 is an example. If random bits are
+cached, then a call to request random bits can somtimes retun those
+bits synchronously.  If the random number generator needs to engage
+the underlying AES engine, then the random bits have to be
+asyncronous. As Rule 1 goes into, even operations that *could* be
+synchronous should have a callback that executes asynchronously.
+
+Having a conditional synchronous operation and an asynchronous backup
+is a poor solution. While it might seem to make the synchronous cases
```

Comment:
It might be worth mentioning that this _might_ be worth it if the operation is synchronous most of the time, and for efficiency reasons it is undesirable to always have to issue a deferred call? I feel like this might be justified in some cases.

Reply:
It definitely is. But as the introduction says, 

"While these are design rules, they are not sarosanct. There are reasons or 
edge cases why a particular HIL might need to break
one (or more) of them. In such cases, be sure to
understand the reasoning behind the rule; if those considerations
don't apply in your use case, then it might be OK to break it. But it's 
important to realize the exception is true for *all*
implementations of the HIL, not just yours; a HIL is intended to be a
general, reusable API, not a specific implementation."

Reply:
Sure. I suppose I was still mentally applying the statement of the previous section ("This rule, however, is more precise and strict.") here. In general this rule does make sense.